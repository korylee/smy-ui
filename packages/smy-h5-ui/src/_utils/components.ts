import type { VueConstructor } from 'vue'

export declare class SmyComponent {
  static name: string
  static install(app: VueConstructor): void
}

function registerComponent(app: VueConstructor, name: string, component: any) {
  const registered = app.component(name)
  if (!registered) {
    app.component(name, component)
  }
}

export function withInstall(component: any) {
  function install(app) {
    const { name, alias } = component
    registerComponent(app, name, component)
    if (alias) {
      alias.forEach((alia) => {
        registerComponent(app, alia, component)
      })
    }
  }
  component.install = install
  return install
}
