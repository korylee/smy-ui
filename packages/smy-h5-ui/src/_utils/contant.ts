export const NAMESPACE = 'smy'

export const BORDER = `${NAMESPACE}--hairline`

export const BORDER_TOP = `${BORDER}-top`
export const BORDER_LEFT = `${BORDER}-left`
export const BORDER_BOTTOM = `${BORDER}-bottom`

export const HAPTICS_FEEDBACK = `${NAMESPACE}--haptics-feedback`

export const ELLIPSIS = `${NAMESPACE}--ellipsis`
