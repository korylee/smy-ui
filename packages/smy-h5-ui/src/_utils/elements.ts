import { kebabCase } from './shared'

export function formatStyleVars(styleVars: Record<string, string>) {
  return Object.entries(styleVars).reduce((styles, [key, value]) => {
    const cssVar = key.startsWith('--') ? key : `--${kebabCase(key)}`
    styles[cssVar] = value
    return styles
  }, {})
}
