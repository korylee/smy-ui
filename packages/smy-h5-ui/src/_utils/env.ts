export const IN_BROWSER = typeof window !== 'undefined'
export const SUPPORT_INTERSECTION = IN_BROWSER && 'IntersectionObserver' in window
