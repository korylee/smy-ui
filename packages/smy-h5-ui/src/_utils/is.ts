export declare interface Func<T = any, R = T> {
  (...args: T[]): R
}

export const isWindow = (val: unknown): val is Window => val === window

const createIsType =
  <T>(type: string) =>
  (val: unknown): val is T =>
    Object.prototype.toString.call(val) === `[object ${type}]`

export const isType = <T>(type: string, val: unknown): val is T => createIsType<T>(type)(val)

export const isNil = (val: unknown): val is null | undefined => val == null

export const isBool = (val: unknown): val is boolean => typeof val === 'boolean'

export const isString = (val: unknown): val is string => typeof val === 'string'

export const isNumber = (val: unknown): val is number => typeof val === 'number'

export const isFunction = (val: unknown): val is Func => typeof val === 'function'

export const isPlainObject = createIsType<Record<string, any>>('Object')

export const isRegExp = createIsType<RegExp>('RegExp')

export const isDate = createIsType<Date>('Date')

export const isObject = (val: unknown): val is object => typeof val === 'object' && val !== null

export const isArray = Array.isArray

export const isRem = (str: string): boolean => isString(str) && str.endsWith('rem')

export const isPx = (str: string): boolean => isString(str) && str.endsWith('px')

export const isVh = (str: string): boolean => isString(str) && str.endsWith('vh')

export const isVw = (str: string): boolean => isString(str) && str.endsWith('vw')

export const isNumString = (str: unknown): boolean => isString(str) && /^\d+$/.test(str)
