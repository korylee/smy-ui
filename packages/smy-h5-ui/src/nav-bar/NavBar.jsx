import { BORDER_BOTTOM, ELLIPSIS, HAPTICS_FEEDBACK } from '../_utils/contant'
import { createNamespace } from '../_utils/vue/create'
import { props } from './props'
import ArrowLeft from '@smy-h5/icons/dist/es/ArrowLeft'
import SmyIcon from '../icon'
import { useWindowSize } from '../_utils/composable/useWindowSize'
import { getRect } from '@smy-h5/shared'
import { defineComponent, h, nextTick, onMounted, ref, watch } from 'vue'

import '../_styles/common.less'
import './navBar.less'

const [name, bem] = createNamespace('nav-bar')

export default defineComponent({
  name,
  props,
  setup(props, { slots, emit }) {
    const root = ref()
    const placeholder = ref()
    const setHeight = () => {
      if (!root.value || !placeholder.value) {
        return
      }
      const { height } = getRect(root.value)
      placeholder.value.style.height = height + 'px'
    }
    onMounted(() => {
      nextTick(setHeight)
      for (let i = 0; i < 3; i++) {
        setTimeout(setHeight, 100 * i)
      }
      const windowSize = useWindowSize()
      watch(() => windowSize, setHeight, { deep: true })
    })

    const onClickLeft = (event) => {
      if (!props.leftDisabled) {
        emit('click-left', event)
      }
    }
    const onClickRight = (event) => {
      if (!props.rightDisabled) {
        emit('click-right', event)
      }
    }

    const renderLeft = () => {
      const leftSlot = slots.left?.()
      if (leftSlot) {
        return leftSlot
      }
      const { leftArrow, leftText } = props
      if (!leftArrow && !leftText) {
        return null
      }
      return [
        leftArrow && <SmyIcon name={leftArrow === true ? ArrowLeft : leftArrow} class={bem('arrow')} />,
        leftText && <span class={bem('text')}>{leftText}</span>,
      ]
    }

    const renderRight = () => {
      const rightSlot = slots.right?.()
      if (rightSlot) {
        return rightSlot
      }
      const { rightText } = props
      if (!rightText) {
        return null
      }
      return <span class={bem('text')}>{rightText}</span>
    }

    const renderContent = () => {
      const { fixed, leftDisabled, clickable, rightDisabled } = props
      const leftVNode = renderLeft()
      const rightVNode = renderRight()
      return (
        <div
          ref="root"
          class={bem({ fixed, '$--safe-area-top': props.safeAreaInsetTop, [BORDER_BOTTOM]: props.border })}
        >
          <div class={bem('content')}>
            {leftVNode &&
              h(
                'div',
                {
                  class: bem('left', { disabled: leftDisabled, [HAPTICS_FEEDBACK]: clickable && !leftDisabled }),
                  on: { click: onClickLeft },
                },
                [leftVNode],
              )}
            <div class={bem('title', ELLIPSIS)}>{slots.title ? slots.title() : props.title}</div>
            {rightVNode &&
              h(
                'div',
                {
                  class: bem('right', { disabled: rightDisabled, [HAPTICS_FEEDBACK]: clickable && !rightDisabled }),
                  on: { click: onClickRight },
                },
                [rightVNode],
              )}
          </div>
        </div>
      )
    }

    return () => {
      if (props.fixed && props.placeholder) {
        return (
          <div ref="placeholder" class={bem('placeholder')}>
            {renderContent()}
          </div>
        )
      }

      return renderContent()
    }
  },
})
