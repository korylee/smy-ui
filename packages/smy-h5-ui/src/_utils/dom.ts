import { IN_BROWSER } from './env'
import { isFunction, isNumString, isNumber, isRem, isPx, isVw, isVh, isWindow, isNumeric } from './is'

export function getAllParentScroller(el: HTMLElement): Array<HTMLElement | Window> {
  const allParentScroller: Array<HTMLElement | Window> = []
  for (let element = getParentScroller(el); element !== window; element = getParentScroller(element as HTMLElement)) {
    allParentScroller.push(element)
  }
  return allParentScroller
}

type ScrollerElement = HTMLElement | Window

export function getParentScroller(el: HTMLElement, root: ScrollerElement = window): HTMLElement | Window {
  let element = el
  while (element && element.parentNode && el !== root) {
    element = element.parentNode as HTMLElement
    if (element === document.body || element === document.documentElement) {
      break
    }
    const scrollRE = /(scroll|auto)/
    const { overflowY, overflow } = window.getComputedStyle(element)
    if (scrollRE.test(overflowY) || scrollRE.test(overflow)) {
      return element
    }
  }
  return root
}

export function requestAnimationFrame(fn: FrameRequestCallback): number {
  return IN_BROWSER ? globalThis.requestAnimationFrame(fn) : setTimeout(fn, 1000 / 60)
}

export function cancelAnimationFrame(handle: number): void {
  return IN_BROWSER ? globalThis.cancelAnimationFrame(handle) : clearTimeout(handle)
}

export const doubleRaf = (cb?: () => void, ctx?: any): Promise<void> => {
  const promise = new Promise<void>((resolve) => {
    requestAnimationFrame(() => requestAnimationFrame(() => resolve()))
  })
  if (!isFunction(cb)) return promise
  return promise.then(() => cb.call(ctx))
}

/**
 * 检查当前dom是否在视口
 */
export function inViewport(el: HTMLElement): Promise<boolean> {
  return doubleRaf().then(() => {
    const { top, bottom, left, right } = getRect(el)
    const { innerWidth, innerHeight } = window
    const xInViewport = left <= innerWidth && right >= 0
    const yInViewport = top <= innerHeight && bottom >= 0

    return xInViewport && yInViewport
  })
}

export const getScrollTopRoot = () =>
  window.pageYOffset || document.documentElement.scrollTop || document.body.scrollTop || 0

export function convertToUnit(str: string | number | null | undefined, unit = 'px'): string | undefined {
  if (str == null || str === '') return undefined
  if (isNumeric(str)) return `${Number(str)}${unit}`
  if (isNaN(+str)) return str
  if (!isFinite(+str)) return undefined
  return `${Number(str)}${unit}`
}

let rootFontSize: number

function getRootFontSize() {
  if (rootFontSize) return rootFontSize
  const doc = document.documentElement
  const fontSize = doc.style.fontSize || window.getComputedStyle(doc).fontSize
  rootFontSize = parseFloat(fontSize)
  return rootFontSize
}

export function toPxNum(value: number | string) {
  if (isNumber(value)) return value
  if (isNumString(value)) return +value
  if (isPx(value)) {
    return +value.replace('px', '')
  }
  if (isVw(value)) {
    return (+value.replace('vw', '') * globalThis.innerWidth) / 100
  }
  if (isVh(value)) {
    return (+value.replace('vh', '') * globalThis.innerHeight) / 100
  }
  if (isRem(value)) {
    return +value.replace('rem', '') * getRootFontSize()
  }
  // % and other
  return 0
}

export const makeDomRect = (width = 0, height = 0) =>
  ({
    top: 0,
    left: 0,
    right: width,
    bottom: height,
    width,
    height,
  } as DOMRect)

export function getRect(el: Element | Window | undefined) {
  if (isWindow(el)) {
    const { innerWidth, innerHeight } = el
    return makeDomRect(innerWidth, innerHeight)
  }
  return el?.getBoundingClientRect?.() ?? makeDomRect()
}

export function getTranslate(el: HTMLElement) {
  const { transform } = window.getComputedStyle(el)
  return +transform.slice(transform.lastIndexOf(',') + 2, transform.length - 1)
}
