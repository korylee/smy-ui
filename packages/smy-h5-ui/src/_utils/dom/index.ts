import { isFunction } from '../is'

export function getAllParentScroller(el: HTMLElement): Array<HTMLElement | Window> {
  const allParentScroller: Array<HTMLElement | Window> = []
  let element: HTMLElement | Window = el
  while (element !== window) {
    element = getParentScroller(element as HTMLElement)
    allParentScroller.push(element)
  }
  return allParentScroller
}

type ScrollerElement = HTMLElement | Window

export function getParentScroller(el: HTMLElement, root: ScrollerElement = window): HTMLElement | Window {
  let element = el
  while (element && el !== root) {
    if (!element.parentNode) break
    element = element.parentNode as HTMLElement
    if (element === document.body || element === document.documentElement) {
      break
    }
    const scrollRE = /(scroll|auto)/
    const { overflowY, overflow } = window.getComputedStyle(element)
    if (scrollRE.test(overflowY) || scrollRE.test(overflow)) {
      return element
    }
  }
  return root
}

export const doubleRaf = (cb?: FrameRequestCallback, ctx?: any) => {
  const promise = new Promise((resolve) => {
    window.requestAnimationFrame(() => {
      window.requestAnimationFrame(resolve)
    })
  })
  if (!isFunction(cb)) return promise
  return promise.then(() => cb.call(ctx))
}

/**
 * 检查当前dom是否在视口，getBoundingClientRect这方案终究还是有性能问题，后续看有无必要换方案
 */
export async function inViewport(el: HTMLElement): Promise<boolean> {
  await doubleRaf()
  const { top, bottom, left, right } = el.getBoundingClientRect()
  const { innerWidth, innerHeight } = window
  const xInViewport = left <= innerWidth && right >= 0
  const yInViewport = top <= innerHeight && bottom >= 0

  return xInViewport && yInViewport
}

export const getScrollTopRoot = () =>
  window.pageYOffset || document.documentElement.scrollTop || document.body.scrollTop || 0
