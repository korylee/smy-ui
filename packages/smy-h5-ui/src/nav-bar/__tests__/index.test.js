import { mount } from '@vue/test-utils'
import NavBar from '..'
import { HAPTICS_FEEDBACK } from '../../_utils/contant'
import { delay, mockGetBoundingClientRect } from '../../../jest-utils'

test('should render left slot correctly', () => {
  const wrapper = mount(NavBar, {
    slots: {
      left: '<div class="custom-left">custom left</div>',
    },
  })
  expect(wrapper.find('.smy-nav-bar__left').html()).toMatchSnapshot()
  expect(wrapper.find('.smy-nav-bar__left').classes(HAPTICS_FEEDBACK)).toBeTruthy()
})

test('should render right slot correctly', () => {
  const wrapper = mount(NavBar, {
    slots: {
      right: '<div class="custom-right">custom right</div>',
    },
  })
  expect(wrapper.find('.smy-nav-bar__right').html()).toMatchSnapshot()
  expect(wrapper.find('.smy-nav-bar__right').classes(HAPTICS_FEEDBACK)).toBeTruthy()
})

test('should render title slot correctly', () => {
  const wrapper = mount(NavBar, {
    slots: {
      title: 'Custom Title',
    },
  })
  expect(wrapper.find('.smy-nav-bar__title').html()).toMatchSnapshot()
})

test('should emit click-right event when clicking right text', () => {
  const wrapper = mount(NavBar, {
    propsData: {
      rightText: 'Custom Right Text',
    },
  })
  wrapper.find('.smy-nav-bar__right').trigger('click')
  expect(wrapper.emitted('click-right')).toBeTruthy()
})

test('should emit click-left event when clicking left text', () => {
  const wrapper = mount(NavBar, {
    propsData: {
      leftText: 'Custom Left Text',
    },
  })
  wrapper.find('.smy-nav-bar__left').trigger('click')
  expect(wrapper.emitted('click-left')).toBeTruthy()
})

test('should render placeholder element when using placeholder prop', async () => {
  const restore = mockGetBoundingClientRect({ height: 100 })
  const wrapper = mount(NavBar, {
    propsData: {
      placeholder: true,
      fixed: true,
    },
  })
  await delay(0)
  expect(wrapper.html()).toMatchSnapshot()
  restore()
})

test('should render slots correctly when set clickable to false', () => {
  const wrapper = mount(NavBar, {
    propsData: {
      clickable: false,
    },
    scopedSlots: {
      left: () => 'Custom Left Text',
      right: () => 'Custom Right Text',
    },
  })
  const leftDom = wrapper.find('.smy-nav-bar__left')
  const rightDom = wrapper.find('.smy-nav-bar__right')
  expect(leftDom.html()).toMatchSnapshot()
  expect(rightDom.html()).toMatchSnapshot()
  expect(leftDom.classes(HAPTICS_FEEDBACK)).toBeFalsy()
  expect(rightDom.classes(HAPTICS_FEEDBACK)).toBeFalsy()
})
