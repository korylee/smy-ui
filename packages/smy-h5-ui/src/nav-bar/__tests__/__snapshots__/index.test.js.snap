// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render left slot correctly 1`] = `
"<div class="smy-nav-bar__left smy--haptics-feedback">
  <div class="custom-left">custom left</div>
</div>"
`;

exports[`should render placeholder element when using placeholder prop 1`] = `
"<div class="smy-nav-bar__placeholder" style="height: 100px;">
  <div class="smy-nav-bar smy-nav-bar--fixed smy--hairline-bottom">
    <div class="smy-nav-bar__content">
      <div class="smy-nav-bar__title smy--ellipsis"></div>
    </div>
  </div>
</div>"
`;

exports[`should render right slot correctly 1`] = `
"<div class="smy-nav-bar__right smy--haptics-feedback">
  <div class="custom-right">custom right</div>
</div>"
`;

exports[`should render slots correctly when set clickable to false 1`] = `"<div class="smy-nav-bar__left">Custom Left Text</div>"`;

exports[`should render slots correctly when set clickable to false 2`] = `"<div class="smy-nav-bar__right">Custom Right Text</div>"`;

exports[`should render title slot correctly 1`] = `"<div class="smy-nav-bar__title smy--ellipsis">Custom Title</div>"`;
